File: ExercismFSharp/ExercismFSharp/Tournament.fs
module Tournament

type private Result =
    | Win
    | Draw
    | Loss

type private TeamRecord =
    {
        name : string
        matchesPlayed : int
        wins : int
        draws : int
        losses : int
        points : int
    }

let private parseLine (line : string) =
    let foo =
        match line.Split[|';'|] with
        | [|a; b; "win"|]  -> Some [(a, Result.Win); (b, Result.Loss)]
        | [|a; b; "draw"|] -> Some [(a, Result.Draw); (b, Result.Draw)]
        | [|a; b; "loss"|] -> Some [(a, Result.Loss); (b, Result.Win)]
        | _ -> None
    foo

let tally input =
    let results =
        input
        |> List.choose parseLine
        |> List.concat
        |> List.groupBy fst
        |> List.map (fun (key, value) -> (key, value |> List.map snd))
        |> List.map (fun (name, results) ->
            let mp = results |> List.length
            let w = results |> List.filter (fun r -> r = Result.Win) |> List.length
            let d = results |> List.filter (fun r -> r = Result.Draw) |> List.length
            let l = results |> List.filter (fun r -> r = Result.Loss) |> List.length
            let p = 3 * w + d
            { name = name; matchesPlayed = mp; wins = w; draws = d; losses = l; points = p })
        |> List.sortBy(fun record -> (-1 * record.points, record.name))
        |> List.map (fun record -> (sprintf "%-31s| %2i | %2i | %2i | %2i | %2i" record.name record.matchesPlayed record.wins record.draws record.losses record.points))

    let header = "Team                           | MP |  W |  D |  L |  P"
    
    header::results

File: Exercism/python/tournament/tournament.py
from collections import defaultdict


def tally(rows):
    wins = defaultdict(int)
    draws = defaultdict(int)
    losses = defaultdict(int)
    teams = set()

    for row in rows:
        team1, team2, result = row.split(";")
        teams.add(team1)
        teams.add(team2)

        if result == "win":
            wins[team1] += 1
            losses[team2] += 1
        elif result == "draw":
            draws[team1] += 1
            draws[team2] += 1
        elif result == "loss":
            losses[team1] += 1
            wins[team2] += 1
        else:
            raise ValueError("Invalid result" + row)

    results = []

    for team in teams:
        team_wins = wins[team]
        team_draws = draws[team]
        team_losses = losses[team]
        matches_played = team_wins + team_draws + team_losses
        points = 3 * team_wins + team_draws
        formatted_line = "{:31}|  {} |  {} |  {} |  {} |  {}" \
            .format(team, matches_played, team_wins, team_draws, team_losses, points)

        results.append((points, team, formatted_line))

    result = [
        "Team                           | MP |  W |  D |  L |  P"
    ]

    result.extend([x[2] for x in sorted(results, key=lambda x: (-x[0], x[1]))])

    return result
